web/client/components/Snapshot/leaflet/GrabMap.jsx

```diff
@@ -0,0 +1,124 @@
+    },
+    componentDidUpdate() {
+        if (!this.props.active) {
+            this.props.onStatusChange("DISABLED");
+            if (this.props.snapstate.error) {
+                this.props.onSnapshotError(null);
+            }
+        }else if (!this.mapIsLoading(this.props.layers)) {
+            if (this.props.snapstate.state !== "SHOTING") {
+                this.props.onStatusChange("SHOTING");
+            }
+            let layers = [].slice.call(this.mapDiv.getElementsByClassName("leaflet-tile-pane")[0].getElementsByClassName("leaflet-layer"), 0);
+            let queue = layers.map(function(l) {
+                return html2canvas(l, {
+                        logging: false,
+                        proxy: this.proxy,
+                        useCORS: false,
+                        type: "view" });
+            }, this);
+            let cSnap = this.createSnapshot;
+            Promise.all(queue).then(function(canvases) {
+                let c = canvases.reduce(function(pCanv, canv, idx) {
+                    let l = layers[idx];
+                    let cx = pCanv.getContext("2d");
+                    if (l.style && !isNaN(Number.parseFloat(l.style.opacity))) {
+                        cx.globalAlpha = Number.parseFloat(l.style.opacity);
+                    }else {
+                        cx.globalAlpha = 1;
+                    }
+                    cx.drawImage(canv, 0, 0);
+                    return pCanv;
+                });
+                cSnap(c);
+            });
+            // layers.reduce((promise, layer, idx, array) => {
+            //     return promise.then(function(canvas) {
+            //         console.log(idx, canvas);
+            //         return drawLayer(layer, globCanvas.width, globCanvas.height);
+            //     });
+            // }, Promise.resolve(this.canvas)).then(function(c) {console.log(c.toDataURL()); });
+        }
+    },
+    render() {
+        return null;
+    },
+    layerLoading() {
+        if (this.props.snapstate.state !== "SHOTING") {
+            this.props.onStatusChange("SHOTING");
+        }
+        this.toLoad = (this.toLoad) ? this.toLoad : 0;
+        this.toLoad++;
+    },
+    layerLoad() {
+        this.toLoad--;
+        if (this.toLoad === 0) {
+            let map = this.refs.snapMap.map;
+            if (map) {
+                map.once('postcompose', (e) => this.createSnapshot(e.context.canvas));
+            }
+        }
+    },
+    createSnapshot(canvas) {
+        let imgData = canvas.toDataURL('image/png');
+        if ( imgData ) {
+            let size = (atob(imgData.substr(22)).length) / 1024;
+            this.props.onSnapshotReady(imgData, canvas.width, canvas.height, size);
+            this.props.onStatusChange("READY");
+        }
+    },
+    mapIsLoading(layers) {
+        return layers.some((layer) => layer.loading);
+    }
+});
+
+module.exports = GrabLMap;
```

Comment:
Leaflet preview has many issues. 
* Sometimes the map is not complete, layer missing or not well stacked 
![image](https://cloud.githubusercontent.com/assets/1279510/12888085/96e92108-ce78-11e5-9694-2fd43ed83908.png)

* When swich layer or background, the preview do not update. 
![image](https://cloud.githubusercontent.com/assets/1279510/12888005/23b24098-ce78-11e5-9a7a-f58b7a6502e8.png)
* it is slow